python/pyspark/sql/types.py

```diff
@@ -187,8 +187,11 @@ def needConversion(self):
 
     def toInternal(self, dt):
         if dt is not None:
-            seconds = (calendar.timegm(dt.utctimetuple()) if dt.tzinfo
-                       else time.mktime(dt.timetuple()))
+            seconds = calendar.timegm(dt.utctimetuple())
+            # Avoiding the invalid range of years (100-1899) for mktime in Python < 3
+            if dt.year not in range(100, 1900):
```

Comment:
Thanks! Indeed, I thought about it but creating PRs too late in the afternoon dizzies my optimisation logic. Great catch and check, although I think `100 >= dt.year > 1900` is an and condition, so it's never only valid for years larger than 1900. I changed it for an or (the < 100 and the > 1899 parts)